<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Dynamic eQTL Calling</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.1/htmlwidgets.js"></script>
<script src="site_libs/plotly-binding-4.9.2.1/plotly.js"></script>
<script src="site_libs/typedarray-0.1/typedarray.min.js"></script>
<link href="site_libs/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.1.0.1/js/crosstalk.min.js"></script>
<link href="site_libs/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="site_libs/plotly-main-1.52.2/plotly-latest.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">sc-dynamic-eqtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmp448/sc-dynamic-eqtl">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Dynamic eQTL Calling</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-01-27
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>sc-dynamic-eqtl/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201027code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201027)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201027code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201027)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjmp448scdynamiceqtltree226d68b1ee3f7876e45e0f9806c0732c37ee3c5etargetblank226d68ba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jmp448/sc-dynamic-eqtl/tree/226d68b1ee3f7876e45e0f9806c0732c37ee3c5e" target="_blank">226d68b</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjmp448scdynamiceqtltree226d68b1ee3f7876e45e0f9806c0732c37ee3c5etargetblank226d68ba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jmp448/sc-dynamic-eqtl/tree/226d68b1ee3f7876e45e0f9806c0732c37ee3c5e" target="_blank">226d68b</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/annotation.nb.html
    Ignored:    analysis/cell-line-pca.nb.html
    Ignored:    analysis/deconvolution.nb.html
    Ignored:    analysis/dynamic-eqtl-2.nb.html
    Ignored:    analysis/dynamic-eqtl.nb.html
    Ignored:    analysis/epigenetics-dynamic.nb.html
    Ignored:    analysis/epigenetics-interaction.nb.html
    Ignored:    analysis/epigenetics.nb.html
    Ignored:    analysis/figure/
    Ignored:    analysis/interaction-qtl-2.nb.html
    Ignored:    analysis/interaction-qtl.nb.html
    Ignored:    analysis/static_eqtl.nb.html
    Ignored:    code/.Rhistory
    Ignored:    data/ben/
    Ignored:    data/bulk_counts.full.tsv
    Ignored:    data/bulk_counts.overlap.tsv
    Ignored:    data/bulk_cpm.full.tsv
    Ignored:    data/bulk_cpm.overlap.tsv
    Ignored:    data/bulk_logtpm.full.tsv
    Ignored:    data/bulk_logtpm.overlap.tsv
    Ignored:    data/bulk_rpkm.full.tsv
    Ignored:    data/bulk_rpkm.overlap.tsv
    Ignored:    data/bulk_tpm.full.tsv
    Ignored:    data/bulk_tpm.overlap.tsv
    Ignored:    data/cibersort/
    Ignored:    data/dynamic/
    Ignored:    data/epigenomes/
    Ignored:    data/gene_locs.filtered.tsv
    Ignored:    data/genotypes.filtered.tsv
    Ignored:    data/gv_pairs.filtered.100k.tsv
    Ignored:    data/gv_pairs.filtered.10k.tsv
    Ignored:    data/gv_pairs.filtered.25k.tsv
    Ignored:    data/gv_pairs.filtered.50k.tsv
    Ignored:    data/gv_pairs.filtered.tsv
    Ignored:    data/pseudobulk-drop2_cpm.col.full.tsv
    Ignored:    data/pseudobulk-drop2_cpm.day.full.tsv
    Ignored:    data/pseudobulk-drop2_logcpm.col.full.tsv
    Ignored:    data/pseudobulk-drop2_logcpm.day.full.tsv
    Ignored:    data/pseudobulk_counts.cmbin.full.tsv
    Ignored:    data/pseudobulk_counts.col.full.tsv
    Ignored:    data/pseudobulk_counts.col.overlap.tsv
    Ignored:    data/pseudobulk_counts.day.full.tsv
    Ignored:    data/pseudobulk_counts.day.overlap.tsv
    Ignored:    data/pseudobulk_counts.epdcbin.full.tsv
    Ignored:    data/pseudobulk_counts.type.full.tsv
    Ignored:    data/pseudobulk_counts.type.overlap.tsv
    Ignored:    data/pseudobulk_cpm.cmbin.full.tsv
    Ignored:    data/pseudobulk_cpm.col.full.tsv
    Ignored:    data/pseudobulk_cpm.col.overlap.tsv
    Ignored:    data/pseudobulk_cpm.day.full.tsv
    Ignored:    data/pseudobulk_cpm.day.overlap.tsv
    Ignored:    data/pseudobulk_cpm.epdcbin.full.tsv
    Ignored:    data/pseudobulk_cpm.type.full.tsv
    Ignored:    data/pseudobulk_cpm.type.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.cmbin.full.tsv
    Ignored:    data/pseudobulk_logcpm.col.full.tsv
    Ignored:    data/pseudobulk_logcpm.col.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.day.full.tsv
    Ignored:    data/pseudobulk_logcpm.day.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.epdcbin.full.tsv
    Ignored:    data/pseudobulk_logcpm.type.full.tsv
    Ignored:    data/pseudobulk_logcpm.type.overlap.tsv
    Ignored:    data/seurat.filtered.rds
    Ignored:    data/seurat.normalized.rds
    Ignored:    data/seurat.processed.rds
    Ignored:    data/seurat/
    Ignored:    data/snp_locs.filtered.tsv
    Ignored:    data/static/
    Ignored:    results/

Untracked files:
    Untracked:  analysis/interaction-qtl-2.Rmd
    Untracked:  code/cell_line_pca.R
    Untracked:  code/cibersort.R
    Untracked:  code/correlation.R
    Untracked:  code/create_seurat.R
    Untracked:  code/debug.R
    Untracked:  code/eqtl_dynamic.R
    Untracked:  code/eqtl_interaction.R
    Untracked:  code/eqtl_static.R
    Untracked:  code/helpers.R
    Untracked:  code/list_tests.R
    Untracked:  code/mashr_static.R
    Untracked:  code/mtc_dynamic.R
    Untracked:  code/mtc_interaction.R
    Untracked:  code/mtc_static.R
    Untracked:  code/normalize_seurat.R
    Untracked:  code/permutation_dynamic.R
    Untracked:  code/permutation_mtc_dynamic.R
    Untracked:  code/prediction_clusters.R
    Untracked:  code/prediction_lasso.R
    Untracked:  code/preprocessing.R
    Untracked:  code/process_seurat.R
    Untracked:  code/replication.R
    Untracked:  figs/
    Untracked:  wrappers/

Unstaged changes:
    Modified:   .gitignore
    Deleted:    analysis/cell-deconvolution.Rmd
    Modified:   analysis/cell-line-pca.Rmd
    Deleted:    analysis/cibersort-calibration.Rmd
    Deleted:    analysis/deconvolution-results.Rmd
    Modified:   analysis/dynamic-eqtl.Rmd
    Modified:   analysis/epigenetics-dynamic.Rmd
    Modified:   analysis/static_eqtl.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/dynamic-eqtl.Rmd</code>) and HTML (<code>docs/dynamic-eqtl.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jmp448/sc-dynamic-eqtl/fdc814092dbd39b17c4a24d8eed0bc4907fde90c/docs/dynamic-eqtl.html" target="_blank">fdc8140</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-25
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jmp448/sc-dynamic-eqtl/72c104151cde03a660e78fe7a80bd09b5593e416/docs/dynamic-eqtl.html" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/c6741b19442d44c5bfd8859aec9ed6522ba75cda/analysis/dynamic-eqtl.Rmd" target="_blank">c6741b1</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
<td>
wflow_publish(“analysis/dynamic-eqtl.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jmp448/sc-dynamic-eqtl/b4aafa83ce34c75d6f9d36c37d37e5285487166b/docs/dynamic-eqtl.html" target="_blank">b4aafa8</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/35e2c763a7bce4ecdb109dc7d8688a28ba385001/analysis/dynamic-eqtl.Rmd" target="_blank">35e2c76</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-12
</td>
<td>
wflow_publish(“analysis/dynamic-eqtl.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jmp448/sc-dynamic-eqtl/adf49b21f72dfe47e15dddbe857e64f642264b13/docs/dynamic-eqtl.html" target="_blank">adf49b2</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-12
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/3d902426341612fb324c8240a98116271595353e/analysis/dynamic-eqtl.Rmd" target="_blank">3d90242</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-12
</td>
<td>
wflow_publish(“analysis/dynamic-eqtl.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="model-overview" class="section level2">
<h2>Model Overview</h2>
<p>Our dynamic eQTL calling framework is based on the following model: <span class="math display">\[{\bf y}_g \sim {\bf K}_{s} +{\bf K}_{cl}+({\bf K}_{cl}*{\bf t}) + {\bf G}_v + ({\bf G}_v*{\bf t})+{\bf t}\]</span> where<br />
<span class="math inline">\({\bf y}_{g}\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing the log-normalized gene expression for gene <span class="math inline">\(g\)</span> for each (individual, time point) sample<br />
<span class="math inline">\({\bf K}_s\)</span> is a <span class="math inline">\(C_s*(nt)\)</span> covariates matrix for <span class="math inline">\(C\)</span> sample covariates (an intercept plus expression PCs) <span class="math inline">\({\bf K}_{cl}\)</span> is a <span class="math inline">\(C_{cl}*(nt)\)</span> covariates matrix of cell-line covariates <span class="math inline">\({\bf G}_v\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing the genotype dosages for each sample<br />
<span class="math inline">\({\bf t}_v\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing the time of collection for each sample<br />
<span class="math inline">\(({\bf K}*{\bf t})\)</span> and <span class="math inline">\(({\bf G}*{\bf t})\)</span> capture the interaction of time with cell line PCs and genotype, respectively</p>
<p>This model is applied to each gene-variant pair. A gene-variant pair is considered a significant dynamic eQTL (dynQTL) if it has a Storey’s q-value <span class="math inline">\(\leq 0.05\)</span> after Bonferroni multiple testing correction.</p>
</div>
<div id="bulk-dynamic-eqtl-calling" class="section level2">
<h2>Bulk Dynamic eQTL Calling</h2>
<p>In order to avoid detecting false positives due to differences in the way cell lines proceed through differentiation, we regress out cell line PCs, as well as expression PCs <img src="figure/dynamic-eqtl.Rmd/compare_detection-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-compare_detection-1">
Past versions of compare_detection-1.png
</button>
</p>
<div id="fig-compare_detection-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/compare_detection-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/dynamic-eqtl.Rmd/binned_detection-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-binned_detection-1">
Past versions of binned_detection-1.png
</button>
</p>
<div id="fig-binned_detection-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/binned_detection-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="pseudobulk-dynamic-eqtl-calling" class="section level2">
<h2>Pseudobulk Dynamic eQTL Calling</h2>
<p>In order to avoid detecting false positives due to differences in the way cell lines proceed through differentiation, we regress out cell line PCs, as well as expression PCs <img src="figure/dynamic-eqtl.Rmd/compare_detection_pb-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-compare_detection_pb-1">
Past versions of compare_detection_pb-1.png
</button>
</p>
<div id="fig-compare_detection_pb-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/compare_detection_pb-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/dynamic-eqtl.Rmd/binned_detection_pb-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-binned_detection_pb-1">
Past versions of binned_detection_pb-1.png
</button>
</p>
<div id="fig-binned_detection_pb-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/binned_detection_pb-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="replication-rates-between-bulk-and-pseudobulk" class="section level2">
<h2>Replication rates between bulk and pseudobulk</h2>
<div id="cell-line-pcs" class="section level3">
<h3>Cell Line PCs</h3>
<p>To see whether these hits are appearing in both datasets, we can look at <span class="math inline">\(\pi_1\)</span> replication rates. First, we look at replication of bulk hits in pseudobulk, compared to a random set of tests, matched for distance to tss. The number of cell line PCs regressed out of bulk is indicated in the column header, pseudobulk in the row header <img src="figure/dynamic-eqtl.Rmd/reprates1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-reprates1-1">
Past versions of reprates1-1.png
</button>
</p>
<div id="fig-reprates1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/reprates1-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It may be more useful to look at replication of pseudobulk hits in bulk, since we would expect a subset of the bulk hits (those corresponding to increased expression among epicardium-like cells) not to be replicated. Here, again, bulk cell line PCs regressed out are shown in column headers, pseudobulk CLPCs regressed are in row headers. <img src="figure/dynamic-eqtl.Rmd/reprates2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-reprates2-1">
Past versions of reprates2-1.png
</button>
</p>
<div id="fig-reprates2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/reprates2-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="sample-pcs" class="section level3">
<h3>Sample PCs</h3>
<p>Now let’s stick to 5 cell line PCs, and see how changing the number of sample PCs will affect this. First, looking at replication of bulk hits in pseudobulk. Columns are sample PCs regressed out of bulk, rows are sample PCs regressed out of pseudobulk. <img src="figure/dynamic-eqtl.Rmd/reprates_samp-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-reprates_samp-1">
Past versions of reprates_samp-1.png
</button>
</p>
<div id="fig-reprates_samp-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/reprates_samp-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now, look at replication of pseudobulk hits in bulk. Again, columns are sample PCs regressed out of bulk, rows are sample PCs regressed out of pseudobulk. <img src="figure/dynamic-eqtl.Rmd/reprates2_samp-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-reprates2_samp-1">
Past versions of reprates2_samp-1.png
</button>
</p>
<div id="fig-reprates2_samp-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/reprates2_samp-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="p-value-comparison-between-experiments" class="section level2">
<h2>P-value comparison between experiments</h2>
<p>Looking across all tests, we see no correlation of p-values <img src="figure/dynamic-eqtl.Rmd/pcomp_scatter-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pcomp_scatter-1">
Past versions of pcomp_scatter-1.png
</button>
</p>
<div id="fig-pcomp_scatter-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/pcomp_scatter-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre><code>[1] 0.05603293</code></pre>
<p>Finally, we can look specifically at those dynamic eQTLs that <em>were</em> significant at <span class="math inline">\(q&lt;=0.05\)</span> in both experiments (mouse over a point to see the snp-gene pair) <div id="htmlwidget-bfc3f274232c9c466624" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-bfc3f274232c9c466624">{"x":{"data":[{"x":[25.6017588570392,23.4644577556426,14.9542685743293,14.7886621449146,11.9966585453401,10.071286355668,9.16841421832114,8.74820478148579,7.92489654003877,7.08687269436458,6.90805485457979,5.99509205227376,5.66546131705509,5.3326703105447,5.04964073413119,4.90966969674892,4.87373181408052,4.76932746896883,4.53773242743913,4.34771750592476,4.24108060761224,4.15190711313401,4.05924658385998,3.77082094007496,3.76652532236252,3.62220659978328,3.51412110363597,3.47265875385664,3.3591872876259,3.2888725480229,3.25003675113135,3.23475456329118,3.00387525044017,2.95871602760029,2.88547206060001,2.87586672030087,2.85688991432769,2.8214910605162,2.75930159241298],"y":[11.3546026496429,16.3859795127663,10.4408676981037,6.18077967489761,3.99833175693935,2.97745318461281,5.80639887702007,4.2010936785221,4.13700971980077,9.13625027864676,3.64414185084508,8.70479765094859,3.64176510479183,5.34178464605301,4.94109211244949,4.48039214093076,4.4342129867775,4.03047386576158,7.9350462548555,6.2641634392836,4.29799585923559,6.69790706252566,7.0075633625007,5.4964131473345,5.33618815367255,3.73606181494135,4.09938566505397,3.41056576243895,4.2632334544566,3.57249659180459,3.39419450383165,3.57941967826986,4.19163994036444,5.98893436341864,4.11588919129039,3.09366744551361,3.95622355313293,5.14897399167744,2.94099602565306],"text":["LDHC--chr11_18412342","FABP2--chr4_119272392","PLA2G2A--chr1_19979653","TMEM176A--chr7_150793288","GJD2--chr15_34751391","DTD1--chr20_18542358","PPIP5K2--chr5_103086881","IGFBP1--chr7_45893070","TGM2--chr20_38142186","WDR60--chr7_158879928","FIGN--chr2_163607546","CENPU--chr4_184649399","CYP2B6--chr19_41012316","ITGA4--chr2_181486016","DYNLT1--chr6_158633395","RNPC3--chr1_103565559","NUP54--chr4_76159678","WDR59--chr16_74904185","ACSM1--chr16_20669907","CCND1--chr11_69653462","IL7R--chr5_35877403","FYN--chr6_111661054","ZNF560--chr19_9435125","HADHA--chr2_26187699","MTCH1--chr6_36968830","DUSP10--chr1_221736946","CCDC74A--chr2_131484131","NEFM--chr8_24917170","CSF1R--chr5_150053759","SRPK1--chr6_35819722","HABP2--chr10_113590838","DDT--chr22_23971817","ZNF483--chr9_111543296","CCL20--chr2_227815864","TMEM71--chr8_132638081","VSNL1--chr2_17515306","GRIA1--chr5_153525446","SSPN--chr12_26122622","ZDHHC17--chr12_76764639"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":37.2602739726027},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[1.61717872918167,26.7438817202705],"tickmode":"array","ticktext":["5","10","15","20","25"],"tickvals":[5,10,15,20,25],"categoryorder":"array","categoryarray":["5","10","15","20","25"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"bulk.log10p","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[2.2687468512974,17.0582286871219],"tickmode":"array","ticktext":["4","8","12","16"],"tickvals":[4,8,12,16],"categoryorder":"array","categoryarray":["4","8","12","16"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"pseudobulk.log10p","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"337cb51768c2":{"x":{},"y":{},"text":{},"type":"scatter"}},"cur_data":"337cb51768c2","visdat":{"337cb51768c2":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script></p>
</div>
<div id="dynqtl-visualization" class="section level2">
<h2>dynQTL Visualization</h2>
<p>Visualize both early and late dynamic eQTLs that were detected in bulk <img src="figure/dynamic-eqtl.Rmd/dynqtl_vis-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dynqtl_vis-1">
Past versions of dynqtl_vis-1.png
</button>
</p>
<div id="fig-dynqtl_vis-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/dynqtl_vis-1.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/dynamic-eqtl.Rmd/dynqtl_vis-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dynqtl_vis-2">
Past versions of dynqtl_vis-2.png
</button>
</p>
<div id="fig-dynqtl_vis-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/dynqtl_vis-2.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/dynamic-eqtl.Rmd/dynqtl_vis-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dynqtl_vis-3">
Past versions of dynqtl_vis-3.png
</button>
</p>
<div id="fig-dynqtl_vis-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/dynqtl_vis-3.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/dynamic-eqtl.Rmd/dynqtl_vis-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dynqtl_vis-4">
Past versions of dynqtl_vis-4.png
</button>
</p>
<div id="fig-dynqtl_vis-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/72c104151cde03a660e78fe7a80bd09b5593e416/docs/figure/dynamic-eqtl.Rmd/dynqtl_vis-4.png" target="_blank">72c1041</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="permutation-analysis" class="section level1">
<h1>Permutation Analysis</h1>
<p>To check whether regressing out sample PCs is leading to an increase in false positives, we ran a permutation test <img src="figure/dynamic-eqtl.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.0 (2020-04-24)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ggpubr_0.4.0    patchwork_1.0.1 plotly_4.9.2.1  qvalue_2.21.0  
 [5] forcats_0.5.0   stringr_1.4.0   dplyr_1.0.2     purrr_0.3.4    
 [9] readr_1.3.1     tidyr_1.1.2     tibble_3.0.3    ggplot2_3.3.2  
[13] tidyverse_1.3.0 workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] fs_1.5.0          lubridate_1.7.9   httr_1.4.2        rprojroot_1.3-2  
 [5] tools_4.0.0       backports_1.1.10  R6_2.4.1          DBI_1.1.0        
 [9] lazyeval_0.2.2    colorspace_1.4-1  withr_2.3.0       tidyselect_1.1.0 
[13] curl_4.3          compiler_4.0.0    git2r_0.27.1      cli_2.0.2        
[17] rvest_0.3.6       xml2_1.3.2        labeling_0.3      scales_1.1.1     
[21] digest_0.6.25     foreign_0.8-78    rmarkdown_2.4     rio_0.5.16       
[25] pkgconfig_2.0.3   htmltools_0.5.0   dbplyr_1.4.4      htmlwidgets_1.5.1
[29] rlang_0.4.7       readxl_1.3.1      rstudioapi_0.11   generics_0.0.2   
[33] farver_2.0.3      jsonlite_1.7.1    crosstalk_1.1.0.1 zip_2.1.1        
[37] car_3.0-10        magrittr_1.5      Rcpp_1.0.5        munsell_0.5.0    
[41] fansi_0.4.1       abind_1.4-5       lifecycle_0.2.0   stringi_1.5.3    
[45] whisker_0.4       yaml_2.2.1        carData_3.0-4     plyr_1.8.6       
[49] grid_4.0.0        blob_1.2.1        promises_1.1.1    crayon_1.3.4     
[53] haven_2.3.1       splines_4.0.0     hms_0.5.3         knitr_1.30       
[57] pillar_1.4.6      ggsignif_0.6.0    reshape2_1.4.4    reprex_0.3.0     
[61] glue_1.4.2        evaluate_0.14     data.table_1.13.0 modelr_0.1.8     
[65] vctrs_0.3.4       httpuv_1.5.4      cellranger_1.1.0  gtable_0.3.0     
[69] assertthat_0.2.1  xfun_0.18         openxlsx_4.2.2    broom_0.7.0      
[73] rstatix_0.6.0     later_1.1.0.1     viridisLite_0.3.0 ellipsis_0.3.1   </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
