<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Interaction eQTLs</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">sc-dynamic-eqtl</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmp448/sc-dynamic-eqtl">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Interaction eQTLs</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-01-25
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sc-dynamic-eqtl/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201027code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201027)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201027code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201027)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjmp448scdynamiceqtltree4f6a4b65035d1f2297ba747675f8edb92c59bcd5targetblank4f6a4b6a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jmp448/sc-dynamic-eqtl/tree/4f6a4b65035d1f2297ba747675f8edb92c59bcd5" target="_blank">4f6a4b6</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjmp448scdynamiceqtltree4f6a4b65035d1f2297ba747675f8edb92c59bcd5targetblank4f6a4b6a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jmp448/sc-dynamic-eqtl/tree/4f6a4b65035d1f2297ba747675f8edb92c59bcd5" target="_blank">4f6a4b6</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/annotation.nb.html
    Ignored:    analysis/cell-line-pca.nb.html
    Ignored:    analysis/deconvolution.nb.html
    Ignored:    analysis/dynamic-eqtl-2.nb.html
    Ignored:    analysis/dynamic-eqtl.nb.html
    Ignored:    analysis/epigenetics-dynamic.nb.html
    Ignored:    analysis/epigenetics-interaction.nb.html
    Ignored:    analysis/epigenetics.nb.html
    Ignored:    analysis/interaction-qtl-2.nb.html
    Ignored:    analysis/interaction-qtl.nb.html
    Ignored:    analysis/static_eqtl.nb.html
    Ignored:    code/.Rhistory
    Ignored:    data/ben/
    Ignored:    data/bulk_counts.full.tsv
    Ignored:    data/bulk_counts.overlap.tsv
    Ignored:    data/bulk_cpm.full.tsv
    Ignored:    data/bulk_cpm.overlap.tsv
    Ignored:    data/bulk_logtpm.full.tsv
    Ignored:    data/bulk_logtpm.overlap.tsv
    Ignored:    data/bulk_rpkm.full.tsv
    Ignored:    data/bulk_rpkm.overlap.tsv
    Ignored:    data/bulk_tpm.full.tsv
    Ignored:    data/bulk_tpm.overlap.tsv
    Ignored:    data/cibersort/
    Ignored:    data/dynamic/
    Ignored:    data/epigenomes/
    Ignored:    data/gene_locs.filtered.tsv
    Ignored:    data/genotypes.filtered.tsv
    Ignored:    data/gv_pairs.filtered.100k.tsv
    Ignored:    data/gv_pairs.filtered.10k.tsv
    Ignored:    data/gv_pairs.filtered.25k.tsv
    Ignored:    data/gv_pairs.filtered.50k.tsv
    Ignored:    data/gv_pairs.filtered.tsv
    Ignored:    data/pseudobulk-drop2_cpm.col.full.tsv
    Ignored:    data/pseudobulk-drop2_cpm.day.full.tsv
    Ignored:    data/pseudobulk-drop2_logcpm.col.full.tsv
    Ignored:    data/pseudobulk-drop2_logcpm.day.full.tsv
    Ignored:    data/pseudobulk_counts.cmbin.full.tsv
    Ignored:    data/pseudobulk_counts.col.full.tsv
    Ignored:    data/pseudobulk_counts.col.overlap.tsv
    Ignored:    data/pseudobulk_counts.day.full.tsv
    Ignored:    data/pseudobulk_counts.day.overlap.tsv
    Ignored:    data/pseudobulk_counts.epdcbin.full.tsv
    Ignored:    data/pseudobulk_counts.type.full.tsv
    Ignored:    data/pseudobulk_counts.type.overlap.tsv
    Ignored:    data/pseudobulk_cpm.cmbin.full.tsv
    Ignored:    data/pseudobulk_cpm.col.full.tsv
    Ignored:    data/pseudobulk_cpm.col.overlap.tsv
    Ignored:    data/pseudobulk_cpm.day.full.tsv
    Ignored:    data/pseudobulk_cpm.day.overlap.tsv
    Ignored:    data/pseudobulk_cpm.epdcbin.full.tsv
    Ignored:    data/pseudobulk_cpm.type.full.tsv
    Ignored:    data/pseudobulk_cpm.type.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.cmbin.full.tsv
    Ignored:    data/pseudobulk_logcpm.col.full.tsv
    Ignored:    data/pseudobulk_logcpm.col.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.day.full.tsv
    Ignored:    data/pseudobulk_logcpm.day.overlap.tsv
    Ignored:    data/pseudobulk_logcpm.epdcbin.full.tsv
    Ignored:    data/pseudobulk_logcpm.type.full.tsv
    Ignored:    data/pseudobulk_logcpm.type.overlap.tsv
    Ignored:    data/seurat/
    Ignored:    data/snp_locs.filtered.tsv
    Ignored:    data/static/
    Ignored:    results/

Untracked files:
    Untracked:  analysis/interaction-qtl-2.Rmd
    Untracked:  code/cell_line_pca.R
    Untracked:  code/cibersort.R
    Untracked:  code/correlation.R
    Untracked:  code/debug.R
    Untracked:  code/eqtl_dynamic.R
    Untracked:  code/eqtl_interaction.R
    Untracked:  code/eqtl_static.R
    Untracked:  code/helpers.R
    Untracked:  code/list_tests.R
    Untracked:  code/mashr_static.R
    Untracked:  code/mtc_dynamic.R
    Untracked:  code/mtc_interaction.R
    Untracked:  code/mtc_static.R
    Untracked:  code/prediction_clusters.R
    Untracked:  code/prediction_lasso.R
    Untracked:  code/preprocessing.R
    Untracked:  code/replication.R
    Untracked:  figs/
    Untracked:  wrappers/

Unstaged changes:
    Modified:   .gitignore
    Deleted:    analysis/cell-deconvolution.Rmd
    Deleted:    analysis/cibersort-calibration.Rmd
    Deleted:    analysis/deconvolution-results.Rmd
    Modified:   analysis/static_eqtl.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/interaction-qtl.Rmd</code>) and HTML (<code>docs/interaction-qtl.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jmp448/sc-dynamic-eqtl/blob/4f6a4b65035d1f2297ba747675f8edb92c59bcd5/analysis/interaction-qtl.Rmd" target="_blank">4f6a4b6</a>
</td>
<td>
jmp448
</td>
<td>
2021-01-25
</td>
<td>
wflow_publish(“analysis/interaction-qtl.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="model-overview" class="section level2">
<h2>Model Overview</h2>
<p>Our cell type interaction eQTL calling framework is based on the following model: <span class="math display">\[{\bf y}_g \sim {\bf K}_{s}+ {\bf K}_{cl}+ {\bf I} + ({\bf K}_{cl}*{\bf I}) + {\bf G}_v+({\bf G}_v*{\bf I})\]</span> where<br />
<span class="math inline">\({\bf y}_{g}\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing the log-normalized gene expression for gene <span class="math inline">\(g\)</span> for each (individual, time point) sample<br />
<span class="math inline">\({\bf K}\)</span> is a <span class="math inline">\(C*(nt)\)</span> covariates matrix for <span class="math inline">\(C\)</span> covariates (an intercept plus expression PCs)<br />
<span class="math inline">\({\bf I}\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing cell type proportion for the cell type of interest <span class="math inline">\({\bf G}_v\)</span> is a <span class="math inline">\(1*(nt)\)</span> array containing the genotype dosages for each sample<br />
<span class="math inline">\(({\bf I}*{\bf G}_v)\)</span> captures the interaction of cell type proportion with genotype</p>
<p>This model is applied to each gene-variant pair, in each cell type that was annotated in this dataset (iPSC, mesoderm, cardiac mesoderm, cardiac progenitor, cardiomyocyte, epicardium). A gene-variant pair is considered a significant cell-type interaction eQTL (ieQTL) if it has a Storey’s <span class="math inline">\(q \leq 0.05\)</span> after Bonferroni multiple testing correction.</p>
</div>
<div id="impact-of-pc-regression-on-detection-rates" class="section level2">
<h2>Impact of PC Regression on Detection Rates</h2>
<p><img src="figure/interaction-qtl.Rmd/sighits-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="distribution-of-p-values-for-cm-ieqtls" class="section level2">
<h2>Distribution of P-values for CM-ieQTLs</h2>
<p>As above, we see that the distribution of test statistics changes quite dramatically as a result of both cell line and expression PC regression. In general, it appears that regressing out cell line PCs reduces the number of significant hits, regressing out sample PCs increases it. <img src="figure/interaction-qtl.Rmd/pvals-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="replication-of-bulk-interaction-eqtls-in-pseudobulk-dynamic-eqtl-tests" class="section level2">
<h2>Replication of bulk interaction eQTLs in pseudobulk dynamic eQTL tests</h2>
<div id="cell-line-pcs" class="section level3">
<h3>Cell line PCs</h3>
<p>The replication rates of bulk CM ieQTLs are pretty low in CM-pseudotime dynamic eQTLs, though higher than we would expect for a background set of tests matched for distance to TSS. Here, columns show the number of cell line PCs regressed out in the interaction eQTL model (bulk data), and the rows correspond to the number of cell line PCs regressed out in the dynamic eQTL model (pseudobulk data, aggregated by pseudotime bin) <img src="figure/interaction-qtl.Rmd/reprates_clPC-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="sample-pcs" class="section level3">
<h3>Sample PCs</h3>
<p>Now we look at the impact of regressing out sample PCs on replication rates, fixing the number of cell line PCs at 5. The columns show the number of sample PCs regressed out of the ieQTL model, and the rows show the number regressed out of the dynamic eQTL model <img src="figure/interaction-qtl.Rmd/reprates_sampPC-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="bulk-ieqtl-dynamic-eqtl-replication" class="section level3">
<h3>Bulk ieQTL-dynamic eQTL replication</h3>
<p>It’s important to keep in mind that changes in cell type proportion will not strictly correlate with changes in pseudotime (or real time). To orient our expectations on how much these two tests could agree we can look at the replication of bulk hits in bulk. Once again, sample PCs regressed out of ieQTL model in columns, dynamic eQTL model in rows (in this case, the same dataset is used for both models). <img src="figure/interaction-qtl.Rmd/reprates_bulk-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="cell-type-ieqtl-visualization" class="section level2">
<h2>Cell Type ieQTL Visualization</h2>
<p>Last, we take a look at these supposed cell type interaction eQTLs and how their effects are detected across the time course (or, as in lower plots, pseudotime-course) <img src="figure/interaction-qtl.Rmd/ieqtl_vis-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/interaction-qtl.Rmd/ieqtl_vis-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/interaction-qtl.Rmd/ieqtl_vis-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/interaction-qtl.Rmd/ieqtl_vis-4.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.0 (2020-04-24)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] patchwork_1.0.1 plotly_4.9.2.1  qvalue_2.21.0   forcats_0.5.0  
 [5] stringr_1.4.0   dplyr_1.0.2     purrr_0.3.4     readr_1.3.1    
 [9] tidyr_1.1.2     tibble_3.0.3    ggplot2_3.3.2   tidyverse_1.3.0
[13] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        lubridate_1.7.9   assertthat_0.2.1  rprojroot_1.3-2  
 [5] digest_0.6.25     R6_2.4.1          cellranger_1.1.0  plyr_1.8.6       
 [9] backports_1.1.10  reprex_0.3.0      evaluate_0.14     httr_1.4.2       
[13] pillar_1.4.6      rlang_0.4.7       lazyeval_0.2.2    readxl_1.3.1     
[17] rstudioapi_0.11   data.table_1.13.0 whisker_0.4       blob_1.2.1       
[21] rmarkdown_2.4     labeling_0.3      splines_4.0.0     htmlwidgets_1.5.1
[25] munsell_0.5.0     broom_0.7.0       compiler_4.0.0    httpuv_1.5.4     
[29] modelr_0.1.8      xfun_0.18         pkgconfig_2.0.3   htmltools_0.5.0  
[33] tidyselect_1.1.0  viridisLite_0.3.0 fansi_0.4.1       crayon_1.3.4     
[37] dbplyr_1.4.4      withr_2.3.0       later_1.1.0.1     grid_4.0.0       
[41] jsonlite_1.7.1    gtable_0.3.0      lifecycle_0.2.0   DBI_1.1.0        
[45] git2r_0.27.1      magrittr_1.5      scales_1.1.1      cli_2.0.2        
[49] stringi_1.5.3     farver_2.0.3      reshape2_1.4.4    fs_1.5.0         
[53] promises_1.1.1    xml2_1.3.2        ellipsis_0.3.1    generics_0.0.2   
[57] vctrs_0.3.4       tools_4.0.0       glue_1.4.2        hms_0.5.3        
[61] yaml_2.2.1        colorspace_1.4-1  rvest_0.3.6       knitr_1.30       
[65] haven_2.3.1      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
