---
title: "Dynamic eQTL Calling"
output: html_notebook
---

```{r, include=FALSE}
library(tidyverse)
library(qvalue)
library(plotly)
library(patchwork)
```

## Model Overview
Our dynamic eQTL calling framework is based on the following model:
$${\bf y}_g \sim {\bf K} + {\bf (K*t)} + {\bf G}_v +  ({\bf G}_v*{\bf t})+{\bf t}$$
where  
${\bf y}_{g}$ is a $1*(nt)$ array containing the log-normalized gene expression for gene $g$ for each (individual, time point) sample  
${\bf K}$ is a $C*(nt)$ covariates matrix for $C$ covariates (an intercept plus cell line PCs)  
${\bf G}_v$ is a $1*(nt)$ array containing the genotype dosages for each sample   
${\bf t}_v$ is a $1*(nt)$ array containing the time of collection for each sample  
$({\bf K}*{\bf t})$ and $({\bf G}*{\bf t})$ capture the interaction of time with cell line PCs and genotype, respectively  
  
This model is applied to each gene-variant pair. A gene-variant pair is considered a significant dynamic eQTL (dynQTL) if it has a Storey's q-value $\leq 0.1$ after Bonferroni multiple testing correction.

## dynQTL Visualization 
```{r warning=FALSE, message=FALSE, echo=FALSE}
nqtl <- 5
dynqtl <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-3clpcs.tophits.tsv") %>%
  select(gene, snp, bonf.p.unadj, qval) %>% 
  filter(qval<=0.1) %>% 
  arrange(bonf.p.unadj) %>% 
  slice_head(n=nqtl)
bulk_expr_all <- read_tsv("data/bulk_logtpm.full.tsv")
pseudobulk_expr_all <- read_tsv("data/pseudobulk_logcpm.day.full.tsv")
geno_all <- read_tsv("data/genotypes.filtered.tsv")

for (i in seq(1, nqtl)) {
  egene <- dynqtl$gene[i]
  evar <- dynqtl$snp[i]
  bulk_expr <- bulk_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logtpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="bulk")
  pseudobulk_expr <- pseudobulk_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logtpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="pseudobulk")
  expr <- bind_rows(bulk_expr, pseudobulk_expr)
  geno <- geno_all %>% filter(snp==evar) %>%
    column_to_rownames("snp") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("ind", "genotype")) %>%
    mutate(ind=str_replace(ind, "NA", ""))
  dynvis <- left_join(expr, geno, by="ind") %>%
    mutate(logtpm=as.numeric(logtpm)) %>%
    mutate(genotype=factor(genotype, levels=c("0","1","2"))) %>%
    mutate(ind=factor(ind)) %>%
    mutate(day=factor(day, levels=seq(0, 15)))
  p1<-ggplot(filter(dynvis, exp=="bulk"), aes(x=day, y=logtpm, fill=genotype)) + 
    facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(evar) +
    theme_classic()
  p2 <- ggplot(filter(dynvis, exp=="pseudobulk"), aes(x=day, y=logtpm, fill=genotype)) + 
    facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(evar) +
    theme_classic()
  print(p1 + p2 + plot_layout(ncol=1))
}
```

## Impact of varying the number of cell line PCs
In order to avoid detecting false positives due to differences in the way cell lines proceed through differentiation, we regress out cell line PCs. Here we evaluate the impact of regressing out different numbers of PCs and compare the number of dynamic eQTLs detected in bulk compared to pseudobulk
```{r, warning=FALSE, message=FALSE, echo=FALSE}
for (dataset in c("bulk", "pseudobulk-drop2")) {
  for (npc in seq(0, 5)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/", dataset, "/day/50k-", npc, "clpcs.tophits.tsv")) %>% filter(lfsr<=0.05) %>% nrow
    if ((dataset=="bulk") & (npc==0)) {
      sighits <- tibble("exp"=dataset, "pcs"=npc, "hits"=nhits)
    } else {
      sighits <- bind_rows(sighits, tibble("exp"=dataset, "pcs"=npc, "hits"=nhits))
    }
  }
}

ggplot(sighits, aes(x=pcs, y=hits)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(exp)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

There is a huge difference here between bulk and pseudobulk detection rates, and we can see by subsetting the bulk data to only 7 days (similar to what we have in pseudobulk) could be responsible for a large part of it
```{r, warning=FALSE, message=FALSE, echo=FALSE}
for (dataset in c("bulk7", "pseudobulk-drop2")) {
  for (npc in seq(0, 5)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/", dataset, "/day/50k-", npc, "clpcs.tophits.tsv")) %>% filter(lfsr<=0.05) %>% nrow
    if ((dataset=="bulk7") & (npc==0)) {
      sighits <- tibble("exp"=dataset, "pcs"=npc, "hits"=nhits)
    } else {
      sighits <- bind_rows(sighits, tibble("exp"=dataset, "pcs"=npc, "hits"=nhits))
    }
  }
}

ggplot(sighits, aes(x=pcs, y=hits)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(exp)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

We see that we can bridge the gap by binning pseudobulk by pseudotime bin rather than day, effectively creating more time points
```{r, warning=FALSE, message=FALSE, echo=FALSE}
for (dataset in c("bulk", "pseudobulk")) {
  if (dataset == "bulk") {
    agg <- "day"
  } else {
    agg <- "cmbin"
  }
  for (npc in seq(0, 5)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/", dataset, "/", agg, "/50k-", npc, "clpcs.tophits.tsv")) %>% filter(lfsr<=0.05) %>% nrow
    if ((dataset=="bulk") & (npc==0)) {
      sighits <- tibble("exp"=dataset, "pcs"=npc, "hits"=nhits)
    } else {
      sighits <- bind_rows(sighits, tibble("exp"=dataset, "pcs"=npc, "hits"=nhits))
    }
  }
}

ggplot(sighits, aes(x=pcs, y=hits)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(exp)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

An alternative testing strategy would be to explicitly perform multiple testing correction at the gene level, followed by FDR correction with only the top hit per gene. We see that this does not lead to a notable change in the number of significant dynamic eGenes detected (except in pseudobulk without cell line PC regression, but these are likely to be confounded.
```{r, warning=FALSE, message=FALSE, echo=FALSE}
for (dataset in c("bulk", "pseudobulk")) {
  if (dataset == "bulk") {
    agg <- "day"
  } else {
    agg <- "cmbin"
  }
  for (npc in seq(0, 5)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/", dataset, "/", agg, "/50k-", npc, "clpcs.tophits.tsv")) %>% filter(qval<=0.05) %>% nrow
    if ((dataset=="bulk") & (npc==0)) {
      sighits <- tibble("exp"=dataset, "pcs"=npc, "hits"=nhits)
    } else {
      sighits <- bind_rows(sighits, tibble("exp"=dataset, "pcs"=npc, "hits"=nhits))
    }
  }
}

ggplot(sighits, aes(x=pcs, y=hits)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(exp)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

## Replication rates between bulk and pseudobulk
To see whether these hits are appearing in both datasets, we can look at $\pi_1$ replication rates. First, we look at replication of bulk hits in pseudobulk, compared to a random set of variants. We see that regressing out 3 cell line PCs maximizes the replication rate.
```{r, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs"=numeric())
for (npc in seq(0, 5)) {
  bulk.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", npc, "clpcs.tophits.tsv")) %>%
    filter(qval<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  
  pseudobulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc, "clpcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% bulk.hits$gv) %>%
    .$p.adj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="bulk_hits",
                                           "pi1"=pseudobulk.pi1, 
                                           "npcs"=npc))
  
  random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc, "clpcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    sample_n(nrow(bulk.hits)) %>%
    .$p.adj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="random",
                                           "pi1"=random.pi1, 
                                           "npcs"=npc))
}

ggplot(rep.rates, aes(x=npcs, y=pi1)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(variants))
```

Reversing this, we now look at replication of pseudobulk hits in bulk.
```{r, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs"=numeric())
for (npc in seq(0, 5)) {
  pseudobulk.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc, "clpcs.tophits.tsv")) %>%
    filter(qval<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  
  bulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", npc, "clpcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% pseudobulk.hits$gv) %>%
    .$p.adj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="bulk_hits",
                                           "pi1"=bulk.pi1, 
                                           "npcs"=npc))
  
  random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", npc, "clpcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    sample_n(nrow(pseudobulk.hits)) %>%
    .$p.adj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="random",
                                           "pi1"=random.pi1, 
                                           "npcs"=npc))
}

ggplot(rep.rates, aes(x=npcs, y=pi1)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(variants))
```

### Comparing p values between samples
Looking across all tests, there does not appear to be any correlation (as expected, considering most tests are null)
```{r, warning=FALSE, message=FALSE, echo=FALSE}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(bulk.log10p=-log10(p.adj)) %>%
  select(gv, bulk.log10p)
pseudobulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(pseudobulk.log10p=-log10(p.adj)) %>%
  select(gv, pseudobulk.log10p)
comp <- inner_join(bulk, pseudobulk, by="gv")
ggplot(comp, aes(x=bulk.log10p, y=pseudobulk.log10p, text=gv)) +
  geom_point()
```

Subsetting to only the tests that are significant in at least one of the differentiations, we see that pvalues are negatively correlated between the two experiments
```{r, warning=FALSE, message=FALSE, echo=FALSE}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(bulk.log10p=-log10(p.adj)) %>%
  rename(bulk.lfsr=lfsr) %>%
  select(gv, bulk.log10p, bulk.lfsr)
pseudobulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(pseudobulk.log10p=-log10(p.adj)) %>%
  rename(pseudobulk.lfsr=lfsr) %>%
  select(gv, pseudobulk.log10p, pseudobulk.lfsr)
comp <- inner_join(bulk, pseudobulk, by="gv") %>%
  mutate(min.sig=mapply(min, bulk.lfsr, pseudobulk.lfsr)) %>%
  filter(min.sig<=0.1)
ggplot(comp, aes(x=bulk.log10p, y=pseudobulk.log10p, text=gv)) +
  geom_point() + 
  theme_classic()
cor(comp$pseudobulk.log10p, comp$bulk.log10p)
```

Finally, we can look specifically at those dynamic eQTLs that _were_ significant at $lfsr<=0.1$ in both experiments (mouse over a point to see the snp-gene pair)
```{r, warning=FALSE, message=FALSE, echo=FALSE}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(bulk.log10p=-log10(p.adj)) %>%
  filter(lfsr<=0.1) %>%
  select(gv, bulk.log10p)
pseudobulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-3clpcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(pseudobulk.log10p=-log10(p.adj)) %>%
  filter(lfsr<=0.1) %>%
  select(gv, pseudobulk.log10p)
comp <- inner_join(bulk, pseudobulk, by="gv")
p <- ggplot(comp, aes(x=bulk.log10p, y=pseudobulk.log10p, text=gv)) +
  geom_point() + 
  theme_classic()
ggplotly(p, tooltip="text")
```

