---
title: "Dynamic eQTL Calling"
output: html_notebook
---

```{r libloader, include=FALSE}
library(tidyverse)
library(qvalue)
library(plotly)
library(patchwork)
library(ggpubr)
library(scales)
library(Seurat)
```

## Lineage Subsetting
One of the major advantages of this approach is that we are able to isolate each lineage, and consequently identify lineage-specific dynamic eQTLs. 
```{r}
sc <- readRDS("data/seurat.annotated.rds")
DimPlot(sc, group.by="type")

sc_cm <- readRDS("data/seurat.cm.rds")
cm_endcol <- hue_pal()(7)[1]
FeaturePlot(sc_cm, "pseudotime", cols=c("#F5F5F5", cm_endcol))


sc_cf <- readRDS("data/seurat.cf.rds")
cf_endcol <- hue_pal()(7)[6]
FeaturePlot(sc_cf, "pseudotime", cols=c("#F5F5F5", cf_endcol))
```

## Model Overview
Our dynamic eQTL calling framework is based on the following model:
$${\bf y}_g \sim {\bf K}_{s} +{\bf K}_{cl}+({\bf K}_{cl}*{\bf t}) + {\bf G}_v + ({\bf G}_v*{\bf t})+{\bf t}$$
where  
${\bf y}_{g}$ is a $1*(nt)$ array containing the log-normalized gene expression for gene $g$ for each (individual, time point) sample  
${\bf K}_s$ is a $C_s*(nt)$ covariates matrix for $C$ sample covariates (an intercept plus expression PCs)
${\bf K}_{cl}$ is a $C_{cl}*(nt)$ covariates matrix of cell-line covariates 
${\bf G}_v$ is a $1*(nt)$ array containing the genotype dosages for each sample   
${\bf t}_v$ is a $1*(nt)$ array containing the time of collection for each sample  
$({\bf K}*{\bf t})$ and $({\bf G}*{\bf t})$ capture the interaction of time with cell line PCs and genotype, respectively  
  
This model is applied to each gene-variant pair. A gene-variant pair is considered a significant dynamic eQTL (dynQTL) if it has a Storey's q-value $\leq 0.05$ after Bonferroni multiple testing correction.

## Bulk Dynamic eQTL Calling
In order to avoid detecting false positives due to differences in the way cell lines proceed through differentiation, we regress out cell line PCs, as well as expression PCs
```{r compare_detection, warning=FALSE, message=FALSE, echo=FALSE}
sighits <- tibble("samp.pcs"=numeric(), "cl.pcs"=numeric(), "hits"=numeric())
for (n.samp.pc in c(0, 5, 10, 20)) {
  for (n.cl.pc in c(0, 3, 5, 10)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.tophits.tsv")) %>% 
        filter(qval.unadj<=0.05) %>% 
        nrow
    sighits <- bind_rows(sighits, tibble("samp.pcs"=n.samp.pc, "cl.pcs"=n.cl.pc, "hits"=nhits))
  }
}
sighits$samp.pcs <- factor(sighits$samp.pcs, levels=c(0, 5, 10, 20))
sighits$cl.pcs <- factor(sighits$cl.pcs, levels=c(0, 3, 5, 10))
ggplot(sighits, aes(x=samp.pcs, y=hits, fill=samp.pcs)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(rows=vars(cl.pcs)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

```{r binned_detection, warning=FALSE, message=FALSE, echo=FALSE}
pdist <- tibble("n_samp_pcs"=numeric(), "n_cl_pcs"=numeric(), 
                "rejection_region"=logical(), "p.unadj"=numeric())
for (n.samp.pc in c(0, 5, 10, 20)) {
  for (n.cl.pc in c(0, 3, 5, 10)) {
    siggies <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.tophits.tsv")) %>%
      filter(qval.unadj<=0.05) %>% 
      mutate(gv=paste(gene, snp, sep="--")) %>% 
      .$gv
    alltests <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>% 
      mutate(rejection_region=gv%in%siggies) %>%
      select(p.unadj, rejection_region) %>%
      mutate(n_samp_pcs=n.samp.pc) %>%
      mutate(n_cl_pcs=n.cl.pc)
    pdist <- bind_rows(pdist, alltests)
  }
}
ggplot(pdist, aes(x=p.unadj, fill=rejection_region)) +
  geom_histogram() +
  theme_classic() +
  theme(axis.text.x=element_blank()) +
  facet_grid(rows=vars(n_cl_pcs), cols=vars(n_samp_pcs)) +
  ggtitle("P-value Distribution for Linear Dynamic eQTLs")
```

## Pseudobulk Dynamic eQTL Calling
In order to avoid detecting false positives due to differences in the way cell lines proceed through differentiation, we regress out cell line PCs, as well as expression PCs
```{r compare_detection_pb, warning=FALSE, message=FALSE, echo=FALSE}
sighits <- tibble("samp.pcs"=numeric(), "cl.pcs"=numeric(), "hits"=numeric())
for (n.samp.pc in c(0, 5, 10, 20)) {
  for (n.cl.pc in c(0, 3, 5, 10)) {
    nhits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.tophits.tsv")) %>% 
        filter(qval.unadj<=0.05) %>% 
        nrow
    sighits <- bind_rows(sighits, tibble("samp.pcs"=n.samp.pc, "cl.pcs"=n.cl.pc, "hits"=nhits))
  }
}
sighits$samp.pcs <- factor(sighits$samp.pcs, levels=c(0, 5, 10, 20))
sighits$cl.pcs <- factor(sighits$cl.pcs, levels=c(0, 3, 5, 10))
ggplot(sighits, aes(x=samp.pcs, y=hits, fill=samp.pcs)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(rows=vars(cl.pcs)) +
  theme(axis.text.x=element_text(angle=45)) + 
  ylab("# significant egenes")
```

```{r binned_detection_pb, warning=FALSE, message=FALSE, echo=FALSE}
pdist <- tibble("n_samp_pcs"=numeric(), "n_cl_pcs"=numeric(), 
                "rejection_region"=logical(), "p.unadj"=numeric())
for (n.samp.pc in c(0, 5, 10, 20)) {
  for (n.cl.pc in c(0, 3, 5, 10)) {
    siggies <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.tophits.tsv")) %>%
      filter(qval.unadj<=0.05) %>% 
      mutate(gv=paste(gene, snp, sep="--")) %>% 
      .$gv
    alltests <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", n.cl.pc, "clpcs-", n.samp.pc, "pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>% 
      mutate(rejection_region=gv%in%siggies) %>%
      select(p.unadj, rejection_region) %>%
      mutate(n_samp_pcs=n.samp.pc) %>%
      mutate(n_cl_pcs=n.cl.pc)
    pdist <- bind_rows(pdist, alltests)
  }
}
ggplot(pdist, aes(x=p.unadj, fill=rejection_region)) +
  geom_histogram() +
  theme_classic() +
  theme(axis.text.x=element_blank()) +
  facet_grid(rows=vars(n_cl_pcs), cols=vars(n_samp_pcs)) +
  ggtitle("P-value Distribution for Linear Dynamic eQTLs")
```

## Replication rates between bulk and pseudobulk
### Cell Line PCs
To see whether these hits are appearing in both datasets, we can look at $\pi_1$ replication rates. First, we look at replication of bulk hits in pseudobulk, compared to a random set of tests, matched for distance to tss. The number of cell line PCs regressed out of bulk is indicated in the column header, pseudobulk in the row header
```{r reprates1, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs"=numeric())
all.tests <- read_tsv("data/gv_pairs.filtered.50k.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--"))

get.matched <- function(tests) {
  elim.genes <- str_extract(tests, "[^-]+")
  match.dists <- all.tests %>% 
    filter(gv %in% tests) %>%
    .$dist2tss
  candidates <- all.tests %>%
    filter(!gene %in% elim.genes)
  matched.tests <- bind_rows(map_dfr(match.dists, function(d){slice(candidates, which.min(abs(dist2tss-d)))}))$gv
  matched.tests
}

for (npc.bulk in c(0, 3, 5, 10)) {
  bulk.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk7/day/50k-", npc.bulk, "clpcs-0pcs.tophits.tsv")) %>%
    filter(qval.unadj<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  for (npc.pseudobulk in c(0, 3, 5, 10)) {
    pseudobulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc.pseudobulk, "clpcs-0pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>%
      filter(gv %in% bulk.hits$gv) %>%
      .$p.unadj %>%
      qvalue %>% 
      (function(x){1-x$pi0})
    
    rep.rates <- bind_rows(rep.rates, tibble("variants"="bulk_rejection_region",
                                             "pi1"=pseudobulk.pi1, 
                                             "npcs.bulk"=npc.bulk,
                                             "npcs.pseudobulk"=npc.pseudobulk))
    
    random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc.pseudobulk, "clpcs-5pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>%
      filter(gv %in% get.matched(bulk.hits$gv)) %>%
      .$p.unadj %>%
      qvalue %>% 
      (function(x){1-x$pi0})
    rep.rates <- bind_rows(rep.rates, tibble("variants"="random_tests",
                                             "pi1"=random.pi1, 
                                             "npcs.bulk"=npc.bulk,
                                             "npcs.pseudobulk"=npc.pseudobulk))
  }
}
rep.rates$npcs.bulk <- factor(rep.rates$npcs.bulk, levels=c(0, 3, 5, 10))
rep.rates$npcs.pseudobulk <- factor(rep.rates$npcs.pseudobulk, levels=c(0, 3, 5, 10))
ggplot(rep.rates, aes(x=variants, y=pi1, fill=variants)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(npcs.bulk), rows=vars(npcs.pseudobulk)) + 
  theme(axis.ticks.x=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank())
```

It may be more useful to look at replication of pseudobulk hits in bulk, since we would expect a subset of the bulk hits (those corresponding to increased expression among epicardium-like cells) not to be replicated. Here, again, bulk cell line PCs regressed out are shown in column headers, pseudobulk CLPCs regressed are in row headers.
```{r reprates2, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs.bulk"=numeric(), "npcs.pseudobulk"=numeric())
all.tests <- read_tsv("data/gv_pairs.filtered.50k.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--"))

for (npc.pseudobulk in c(0, 3, 5, 10)) {
  pb.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-", npc.pseudobulk, "clpcs-0pcs.tophits.tsv")) %>%
    filter(qval.unadj<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  for (npc.bulk in c(0, 3, 5, 10)) {
  bulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", npc.bulk, "clpcs-0pcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% pb.hits$gv) %>%
    .$p.unadj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  
  rep.rates <- bind_rows(rep.rates, tibble("variants"="pseudobulk_rejection_region",
                                           "pi1"=bulk.pi1, 
                                           "npcs.bulk"=npc.bulk,
                                           "npcs.pseudobulk"=npc.pseudobulk))
  
  random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-", npc.bulk, "clpcs-0pcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% get.matched(pb.hits$gv)) %>%
    .$p.unadj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="random_tests",
                                           "pi1"=random.pi1, 
                                           "npcs.bulk"=npc.bulk,
                                           "npcs.pseudobulk"=npc.pseudobulk))
  }
}
rep.rates$npcs.bulk <- factor(rep.rates$npcs.bulk, levels=c(0, 3, 5, 10))
rep.rates$npcs.pseudobulk <- factor(rep.rates$npcs.pseudobulk, levels=c(0, 3, 5, 10))
ggplot(rep.rates, aes(x=variants, y=pi1, fill=variants)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(npcs.bulk), rows=vars(npcs.pseudobulk)) + 
  theme(axis.ticks.x=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank())
```

### Sample PCs
Now let's stick to 5 cell line PCs, and see how changing the number of sample PCs will affect this. First, looking at replication of bulk hits in pseudobulk. Columns are sample PCs regressed out of bulk, rows are sample PCs regressed out of pseudobulk. 
```{r reprates_samp, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs.bulk"=numeric(), "npcs.pseudobulk"=numeric())

for (npc.bulk in c(0, 5, 10, 20)) {
  bulk.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-", npc.bulk, "pcs.tophits.tsv")) %>%
    filter(qval.unadj<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  for (npc.pseudobulk in c(0, 5, 10, 20)) {
    pseudobulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-5clpcs-", npc.pseudobulk, "pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>%
      filter(gv %in% bulk.hits$gv) %>%
      .$p.unadj %>%
      qvalue %>% 
      (function(x){1-x$pi0})
    
    rep.rates <- bind_rows(rep.rates, tibble("variants"="bulk_rejection_region",
                                             "pi1"=pseudobulk.pi1, 
                                             "npcs.bulk"=npc.bulk,
                                             "npcs.pseudobulk"=npc.pseudobulk))
    
    random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-5clpcs-", npc.pseudobulk, "pcs.mtc.tsv")) %>%
      mutate(gv=paste(gene, snp, sep="--")) %>%
      filter(gv %in% get.matched(bulk.hits$gv)) %>%
      .$p.unadj %>%
      qvalue %>% 
      (function(x){1-x$pi0})
    rep.rates <- bind_rows(rep.rates, tibble("variants"="random_tests",
                                             "pi1"=random.pi1, 
                                             "npcs.bulk"=npc.bulk,
                                             "npcs.pseudobulk"=npc.pseudobulk))
  }
}
rep.rates$npcs.bulk <- factor(rep.rates$npcs.bulk, levels=c(0, 5, 10, 20))
rep.rates$npcs.pseudobulk <- factor(rep.rates$npcs.pseudobulk, levels=c(0, 5, 10, 20))
ggplot(rep.rates, aes(x=variants, y=pi1, fill=variants)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(npcs.bulk), rows=vars(npcs.pseudobulk)) + 
  theme(axis.ticks.x=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank())
```

Now, look at replication of pseudobulk hits in bulk. Again, columns are sample PCs regressed out of bulk, rows are sample PCs regressed out of pseudobulk.
```{r reprates2_samp, warning = FALSE, message = FALSE, echo=FALSE}
rep.rates <- tibble("variants"=character(), "pi1"=numeric(), "npcs.bulk"=numeric(), "npcs.pseudobulk"=numeric())
all.tests <- read_tsv("data/gv_pairs.filtered.50k.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--"))

for (npc.pseudobulk in c(0, 5, 10, 20)) {
  pb.hits <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-5clpcs-", npc.pseudobulk, "pcs.tophits.tsv")) %>%
    filter(qval.unadj<=0.1) %>% 
    mutate(gv=paste(gene, snp, sep="--")) %>% 
    rename(bulk.p=bonf.p.unadj) %>% 
    select(c(bulk.p, gv))
  for (npc.bulk in c(0, 5, 10, 20)) {
  bulk.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-", npc.bulk, "pcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% pb.hits$gv) %>%
    .$p.unadj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  
  rep.rates <- bind_rows(rep.rates, tibble("variants"="pseudobulk_rejection_region",
                                           "pi1"=bulk.pi1, 
                                           "npcs.bulk"=npc.bulk,
                                           "npcs.pseudobulk"=npc.pseudobulk))
  
  random.pi1 <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-", npc.bulk, "pcs.mtc.tsv")) %>%
    mutate(gv=paste(gene, snp, sep="--")) %>%
    filter(gv %in% get.matched(pb.hits$gv)) %>%
    .$p.unadj %>%
    qvalue %>% 
    (function(x){1-x$pi0})
  rep.rates <- bind_rows(rep.rates, tibble("variants"="random_tests",
                                           "pi1"=random.pi1, 
                                           "npcs.bulk"=npc.bulk,
                                           "npcs.pseudobulk"=npc.pseudobulk))
  }
}
rep.rates$npcs.bulk <- factor(rep.rates$npcs.bulk, levels=c(0, 5, 10, 20))
rep.rates$npcs.pseudobulk <- factor(rep.rates$npcs.pseudobulk, levels=c(0, 5, 10, 20))
ggplot(rep.rates, aes(x=variants, y=pi1, fill=variants)) + 
  geom_bar(position="dodge", stat="identity") +
  facet_grid(cols=vars(npcs.bulk), rows=vars(npcs.pseudobulk)) + 
  theme(axis.ticks.x=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank())
```

## P-value comparison between experiments
Looking across all tests, we see no correlation of p-values
```{r pcomp_scatter, warning=FALSE, message=FALSE, echo=FALSE}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-5pcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(bulk.log10p=-log10(p.unadj)) %>%
  select(gv, bulk.log10p)
pseudobulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-5clpcs-5pcs.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(pseudobulk.log10p=-log10(p.unadj)) %>%
  select(gv, pseudobulk.log10p)
comp <- inner_join(bulk, pseudobulk, by="gv")
ggplot(comp, aes(x=bulk.log10p, y=pseudobulk.log10p, text=gv)) +
  geom_point() +
  stat_cor(aes(x=bulk.log10p, y=pseudobulk.log10p, label = ..r.label..))
cor(comp$bulk.log10p, comp$pseudobulk.log10p)
```

Finally, we can look specifically at those dynamic eQTLs that _were_ significant at $q<=0.05$ in both experiments (mouse over a point to see the snp-gene pair)
```{r consensus_hits, warning=FALSE, message=FALSE, echo=FALSE}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-5pcs.tophits.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(bulk.log10p=-log10(p.unadj)) %>%
  filter(qval.unadj<=0.05) %>%
  select(gv, bulk.log10p)
pseudobulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/cmbin/50k-5clpcs-5pcs.tophits.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  mutate(pseudobulk.log10p=-log10(p.unadj)) %>%
  filter(qval.unadj<=0.05) %>%
  select(gv, pseudobulk.log10p)
comp <- inner_join(bulk, pseudobulk, by="gv")
p <- ggplot(comp, aes(x=bulk.log10p, y=pseudobulk.log10p, text=gv)) +
  geom_point() + 
  theme_classic()
ggplotly(p, tooltip="text")
```

## dynQTL Visualization 
```{r classify_fcn, echo=FALSE, warning=FALSE, message=FALSE}
classify.dynqtl <- function(beta.g, beta.gxt, beta.t, ..., t.low=0, t.high=15, thresh=1) {
  beta.vgt.early = (beta.t*t.low + beta.gxt*0*t.low + beta.g*0) - (beta.t*t.low + beta.gxt*2*t.low + beta.g*2)
  beta.vgt.late = (beta.t*t.high + beta.gxt*0*t.high + beta.g*0) - (beta.t*t.high + beta.gxt*2*t.high + beta.g*2)
  
  if (sign(beta.vgt.early)==sign(beta.vgt.late)) {
    qtl.type = if_else(abs(beta.vgt.early)>=abs(beta.vgt.late), "early", "late")
  } else {
    if ((abs(beta.vgt.early)>=abs(beta.vgt.late)) & (abs(beta.vgt.late)<thresh)) {
      qtl.type = "early"
    } else if ((abs(beta.vgt.early)<abs(beta.vgt.late)) & (abs(beta.vgt.early)<thresh)) {
      qtl.type = "late"
    } else if ((abs(beta.vgt.early)>=thresh) & (abs(beta.vgt.late)>=thresh)) {
      qtl.type = "switch"
    }
  }
}

nbin <- 15
cm.medians <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin", nbin, "/bin_medians.tsv"))
t.low.cm <- cm.medians %>% 
  mutate(bin=str_extract(binind, "[^_]+$")) %>%
  filter(bin == 0) %>%
  .$t %>%
  median
t.high.cm <- cm.medians %>% 
  mutate(bin=str_extract(binind, "[^_]+$")) %>%
  filter(bin == nbin-1) %>%
  .$t %>%
  median

cf.medians <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin", nbin, "/bin_medians.tsv"))
t.low.cf <- cf.medians %>% 
  mutate(bin=str_extract(binind, "[^_]+$")) %>%
  filter(bin == 0) %>%
  .$t %>%
  median
t.high.cf <- cf.medians %>% 
  mutate(bin=str_extract(binind, "[^_]+$")) %>%
  filter(bin == nbin-1) %>%
  .$t %>%
  median
```

Visualize for both CF and CM an early and a late dynamic eQTL
```{r dynqtl_vis, warning=FALSE, message=FALSE, echo=FALSE}
n.qtls <- 5
dynqtl_cm <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin15/25k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>% 
  arrange(p.unadj) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cm, t.high=t.high.cm)) %>%
  group_by(qtl.type) %>%
  slice_head(n=n.qtls) %>%
  mutate(celltype="CM")
dynqtl_cf <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin15/25k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>% 
  arrange(p.unadj) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cf, t.high=t.high.cf)) %>%
  group_by(qtl.type) %>%
  slice_head(n=n.qtls) %>%
  mutate(celltype="CF")
dynqtl <- bind_rows(dynqtl_cm, dynqtl_cf)
cm_expr_all <- read_tsv("data/pseudobulk-cm/bin15/logcpm.tsv")
cf_expr_all <- read_tsv("data/pseudobulk-cf/bin15/logcpm.tsv")
geno_all <- vroom("data/genotypes/genotypes.tsv") %>%
  filter(snp %in% dynqtl$snp)

for (i in seq(1, nrow(dynqtl))) {
  egene <- dynqtl$gene[i]
  evar <- dynqtl$snp[i]
  cm_expr <- cm_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cm")
  cf_expr <- cf_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cf")
  expr <- bind_rows(cm_expr, cf_expr)
  geno <- geno_all %>% filter(snp==evar) %>%
    column_to_rownames("snp") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("ind", "genotype")) %>%
    mutate(ind=str_replace(ind, "NA", ""))
  dynvis <- left_join(expr, geno, by="ind") %>%
    mutate(logtpm=as.numeric(logcpm)) %>%
    mutate(genotype=factor(genotype, levels=c("0","1","2"))) %>%
    mutate(ind=factor(ind)) %>%
    mutate(day=factor(day, levels=seq(0, 15)))
  p1<-ggplot(filter(dynvis, exp=="cm"), aes(x=day, y=logcpm, fill=genotype)) + 
    facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(paste0(evar, "--", dynqtl$qtl.type[i], "--", dynqtl$celltype[i])) +
    theme_classic()
  p2 <- ggplot(filter(dynvis, exp=="cf"), aes(x=day, y=logcpm, fill=genotype)) +     facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(paste0(evar, "--", dynqtl$qtl.type[i], "--", dynqtl$celltype[i])) +
    theme_classic()
  print(p1 + p2 + plot_layout(ncol=1))
}
```

Find a late hit that's shared by the two lineages
```{r}
dynqtl_cm <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin20/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.1) %>% 
  arrange(p.unadj) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cm, t.high=t.high.cm)) %>%
  select(gene, snp, p.unadj, lfsr, qtl.type) %>%
  group_by(qtl.type) %>%
  mutate(celltype="CM")
dynqtl_cf <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin20/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.1) %>% 
  arrange(p.unadj) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cf, t.high=t.high.cf)) %>%
  select(gene, snp, p.unadj, lfsr, qtl.type) %>%
  group_by(qtl.type) %>%
  mutate(celltype="CF")
dynqtl_shared <- inner_join(dynqtl_cm, dynqtl_cf, by=c("gene","snp", "qtl.type")) 
cm_expr_all <- read_tsv("data/pseudobulk-cm/bin20/logcpm.tsv")
cf_expr_all <- read_tsv("data/pseudobulk-cf/bin20/logcpm.tsv")
geno_all <- read_tsv("data/genotypes.filtered.tsv")
for (i in seq(1, nrow(dynqtl_shared))) {
  egene <- dynqtl_shared$gene[i]
  evar <- dynqtl_shared$snp[i]
  cm_expr <- cm_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cm")
  cf_expr <- cf_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cf")
  expr <- bind_rows(cm_expr, cf_expr)
  geno <- geno_all %>% filter(snp==evar) %>%
    column_to_rownames("snp") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("ind", "genotype")) %>%
    mutate(ind=str_replace(ind, "NA", ""))
  dynvis <- left_join(expr, geno, by="ind") %>%
    mutate(logtpm=as.numeric(logcpm)) %>%
    mutate(genotype=factor(genotype, levels=c("0","1","2"))) %>%
    mutate(ind=factor(ind)) %>%
    mutate(day=factor(day, levels=seq(0, 15)))
  p1<-ggplot(filter(dynvis, exp=="cm"), aes(x=day, y=logcpm, fill=genotype)) + 
    facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(paste0(evar, "--", dynqtl_shared$qtl.type[i], "--", dynqtl_shared$celltype[i])) +
    theme_classic()
  p2 <- ggplot(filter(dynvis, exp=="cf"), aes(x=day, y=logcpm, fill=genotype)) +     facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    ggtitle(paste0(evar, "--", dynqtl_shared$qtl.type[i], "--", dynqtl_shared$celltype[i])) +
    theme_classic()
  print(p1 + p2 + plot_layout(ncol=1))
}
```

Identify late dynamic hits that are unique to CM
```{r}
nbin <- 15
dynqtl_cm_uniq <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin", nbin, "/50k-5clpcs-0pcs-notypes.tophits.tsv")) %>%
  filter(lfsr<=0.1) %>% 
  arrange(p.unadj) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cm, t.high=t.high.cm)) %>%
  select(gene, snp, p.unadj, lfsr, qtl.type) %>%
  group_by(qtl.type) %>%
  mutate(celltype="CM") %>%
  mutate(log10p_cm=-log10(p.unadj))
cf_comp <- read_tsv(paste0("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin", nbin, "/50k-5clpcs-0pcs-notypes.mtc.tsv")) %>%
  filter((gene %in% dynqtl_cm_uniq$gene) & (snp %in% dynqtl_cm_uniq$snp)) %>%
  mutate(log10p_cf=-log10(p.unadj)) %>%
  select(gene, snp, log10p_cf) %>%
  right_join(dynqtl_cm_uniq, by=c("snp", "gene"))
cf_comp <- cf_comp %>%
  mutate(logdiff=log10p_cm-log10p_cf) %>%
  arrange(logdiff)
for (i in seq(1, 10)) {
  egene <- cf_comp$gene[i]
  evar <- cf_comp$snp[i]
  cm_expr <- cm_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cm")
  cf_expr <- cf_expr_all %>% filter(gene==!!egene) %>% 
    column_to_rownames("gene") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("sample", "logcpm")) %>%
    mutate(day=str_extract(sample, "[^_]+$")) %>%
    mutate(ind=str_extract(sample, "[^_]+")) %>%
    mutate(exp="cf")
  expr <- bind_rows(cm_expr, cf_expr)
  geno <- geno_all %>% filter(snp==evar) %>%
    column_to_rownames("snp") %>% t %>% 
    as_tibble(rownames="ind") %>% 
    `colnames<-`(c("ind", "genotype")) %>%
    mutate(ind=str_replace(ind, "NA", ""))
  dynvis <- left_join(expr, geno, by="ind") %>%
    mutate(logtpm=as.numeric(logcpm)) %>%
    mutate(genotype=factor(genotype, levels=c("0","1","2"))) %>%
    mutate(ind=factor(ind)) %>%
    mutate(bin=factor(day, levels=seq(0, nbin-1)))
  p1<-ggplot(filter(dynvis, exp=="cm"), aes(x=bin, y=logcpm, fill=genotype)) + 
    facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    theme_classic()
  p2 <- ggplot(filter(dynvis, exp=="cf"), aes(x=bin, y=logcpm, fill=genotype)) +     facet_grid(rows=vars(exp)) +
    geom_boxplot() +
    ylab(egene) +
    theme_classic()
  print(p1 + p2 + plot_layout(ncol=1))
}
```


Replication rate of early and late hits between CF and CM
```{r}
cm_hits <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.1) %>%
  mutate(qtl.type=pmap_chr(., classify.dynqtl, t.low=t.low.cm, t.high=t.high.cm)) %>%
  mutate(gv=paste(gene, snp, sep="--"))
cf_tests <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin/50k-5clpcs-0pcs-notypes.mtc.tsv") %>%
  mutate(gv=paste(gene, snp, sep="--")) %>%
  filter(gv %in% cm_hits$gv) %>%
  left_join(select(cm_hits, gv, qtl.type), by="gv")
ggplot(filter(cf_tests, qtl.type=="early"), aes(x=p.unadj)) + 
  geom_histogram()
ggplot(filter(cf_tests, qtl.type=="late"), aes(x=p.unadj)) + 
  geom_histogram()
ggplot(filter(cf_tests, qtl.type=="switch"), aes(x=p.unadj)) + 
  geom_histogram()
qvalue(filter(cm_hits, qtl.type=="late")$p.unadj)$pi0
```
Compare all metrics - qvalue
```{r}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
bulk7_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk7/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cm_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cf_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cm_bin16 <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin16/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cf_bin16 <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin16/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
methods_used <- c("bulk (day)", 
                                  "pseudobulk (day)", 
                                  "pseudobulk-CM (day)", 
                                  "pseudobulk-CF (day)", 
                                  "pseudobulk-CM (dpt)", 
                                  "pseudobulk-CF (dpt)")
method_comps <- tibble("Method"=factor(methods_used, 
                                       levels=methods_used),
                       "qtlcount"=c(bulk7_day,
                                    pseudobulk_day,
                                    pseudobulk_cm_day, 
                                    pseudobulk_cf_day, 
                                    pseudobulk_cm_bin16,
                                    pseudobulk_cf_bin16),
                       "grp"=c("bulk", "pb", "pb-lin", "pb-lin", "pb-lin-t", "pb-lin-t")
                       )
my_cols <- c(hue_pal()(7)[c(5,4,3,2,1,6)], "#949494")
method_cols <- c("#808080", "#808080", my_cols[5], my_cols[6], "#bf0c00", "#8360f7")
ggplot(method_comps, aes(x=grp, y=qtlcount, fill=Method)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("") +
  ylab("Number of Dynamic eQTLs Detected") +
  theme_classic() +
  scale_fill_manual(values=method_cols) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
```

```{r}
bulk <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
bulk7_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/bulk7/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cm_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cf_day <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/day/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cm_bin16 <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cm/bin16/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cf_bin16 <- read_tsv("results/eqtl_dynamic/linear_dQTL/pseudobulk-cf/bin16/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
methods_used <- c("bulk (day)", 
                                  "pseudobulk (day)", 
                                  "pseudobulk-CM (day)", 
                                  "pseudobulk-CF (day)", 
                                  "pseudobulk-CM (dpt)", 
                                  "pseudobulk-CF (dpt)")
method_comps <- tibble("Method"=factor(methods_used, 
                                       levels=methods_used),
                       "qtlcount"=c(bulk7_day,
                                    pseudobulk_day,
                                    pseudobulk_cm_day, 
                                    pseudobulk_cf_day, 
                                    pseudobulk_cm_bin20,
                                    pseudobulk_cf_bin20),
                       "grp"=c("bulk", "pb", "pb-lin", "pb-lin", "pb-lin-t", "pb-lin-t")
                       )
my_cols <- c(hue_pal()(7)[c(5,4,3,2,1,6)], "#949494")
method_cols <- c("#808080", "#808080", my_cols[5], my_cols[6], "#bf0c00", "#8360f7")
ggplot(method_comps, aes(x=grp, y=qtlcount, fill=Method)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("") +
  ylab("Number of Dynamic eQTLs Detected") +
  theme_classic() +
  scale_fill_manual(values=method_cols) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
```

```{r}
fig3d <- ggplot(method_comps, aes(x=grp, y=qtlcount, fill=Method)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("") +
  ylab("Number of Dynamic eQTLs Detected") +
  theme_classic() +
  scale_fill_manual(values=method_cols) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())

png('figs/fig3/fig3d.png', width=400, height=600)
plot(fig3d)
dev.off()
```


Count nonlinear dynamic eQTLs
```{r}
bulk_nonlinear <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/bulk/day/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cm_nonlinear_bin16 <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/pseudobulk-cm/bin16/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()
pseudobulk_cf_nonlinear_bin16 <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/pseudobulk-cf/bin16/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(qval.unadj<=0.05) %>%
  nrow()

bulk_nonlinear_lfsr <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/bulk/day/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cm_nonlinear_bin16_lfsr <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/pseudobulk-cm/bin16/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
pseudobulk_cf_nonlinear_bin16_lfsr <- read_tsv("results/eqtl_dynamic/nonlinear_dQTL/pseudobulk-cf/bin16/t_test/50k-5clpcs-0pcs-notypes.tophits.tsv") %>%
  filter(lfsr<=0.05) %>%
  nrow()
```



```{r}
t <- read_tsv("data/pseudobulk-cm/bin15/pcs.tsv") %>%
  mutate(day=as.numeric(str_extract(sample, "[^_]+$"))) %>%
  mutate(day=factor(day, levels=seq(0,19)))


ggplot(t, aes(x=PC1, y=PC2, color=day)) + 
  geom_point()
```

